### ISSUE TITLE
sandbox networking doesn't work with rootless podman, hostinet does

### ISSUE THREAD

**Issue by chetan-reddy** [google/gvisor#9398]
### Description

See output below when podman is run with --network=host (this --network flag is independent of --runtime-flag network). The first command shows no interfaces, whereas the second works as expected.

```
$ podman run --runtime=runsc --runtime-flag ignore-cgroups --runtime-flag platform=kvm --runtime-flag network=sandbox --network=host --rm -it docker://alpine:20230901 ip addr
$ podman run --runtime=runsc --runtime-flag ignore-cgroups --runtime-flag platform=kvm --runtime-flag network=host --network=host --rm -it docker://alpine:20230901 ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 0 qlen 1000
    link/loopback 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
    inet 127.0.0.1/8 scope global dynamic
    inet6 ::1/128 scope global dynamic
2: enx... 
```

If I start podman with --network=bridge (the default), both network=sandbox and network=host have working outbound connectivity, but incoming connections don't work with network=sandbox . In output below, the incoming hello message is from running `echo hello | socat TCP4-CONNECT:127.0.0.1:1234 -` in a different terminal
```
$ podman run --runtime=runsc --runtime-flag ignore-cgroups --runtime-flag platform=kvm --runtime-flag network=host --network=bridge -p 127.0.0.1:1234:1234 --rm -it docker://alpine/socat TCP4-LISTEN:1234 STDIO
hello
$ podman run --runtime=runsc --runtime-flag ignore-cgroups --runtime-flag platform=kvm --runtime-flag network=sandbox --network=bridge -p 127.0.0.1:1234:1234 --rm -it docker://alpine/socat TCP4-LISTEN:1234 STDIO
... no messages , just hangs ...
```
 
In both cases (podman --network=host and --network=bridge),  runsc's sandbox networking doesn't work as well as hostinet. Am I missing something?

### runsc version

```shell
$ runsc --version
runsc version release-20230911.0
spec: 1.1.0-rc.1
```

### docker version (if using docker)

```shell
$ podman --version
podman version 4.3.1
```

### uname

Linux mba 6.1.0-12-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.52-1 (2023-09-07) x86_64 GNU/Linux

**Comment by avagin** [comment]
> See output below when podman is run with --network=host (this --network flag is independent of --runtime-flag network). The first command shows no interfaces, whereas the second works as expected.

It is expected. runsc with the sandbox network needs to change a network configuration, but it doesn't have enough permissions for that. Actually, I think you don't want to allow runsc to change the host network configuration even when it is running under the root user... 

> If I start podman with --network=bridge (the default), both network=sandbox and network=host have working outbound connectivity, but incoming connections don't work with network=sandbox .

rootlessport is running in a container network namespace and it is handling mapped ports. If runsc uses the sandbox network, it removes all ip addresses from devices (otherwise the host kernel tcp stack will interfere with the runsc network stack) and it breaks rootlessport:

21679 connect(10, {sa_family=AF_INET, sin_port=htons(8889), sin_addr=inet_addr("10.0.2.100")}, 16 <unfinished ...>
21679 <... connect resumed>)            = -1 ENETUNREACH (Network is unreachable)
21679 write(5, "E\0\0\0{\"Error\":\"dial tcp 10.0.2.100:8889: connect: network is unreachable\"}", 73 <unfinished ...>

# nsenter -n -t 21701 ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: tap0: <BROADCAST,UP,LOWER_UP> mtu 65520 qdisc pfifo_fast state UNKNOWN group default qlen 1000
    link/ether 8a:08:96:de:65:09 brd ff:ff:ff:ff:ff:ff

In summary, if you want to publishes TCP ports, you should run runsc with the host network mode.

ps: podman is not our priority, because we don't use it in our infrastructure.

**Comment by chetan-reddy** [comment]
> rootlessport is running in a container network namespace and it is handling mapped ports. If runsc uses the sandbox network, it removes all ip addresses from devices (otherwise the host kernel tcp stack will interfere with the runsc network stack) and it breaks rootlessport

Thanks for the pointer. It helped me discover that incoming connectivity does work by changing the port handler from rootlessport to slirp4netns. This is done by invoking rootless podman with
 `--network slirp4netns:port_handler=slirp4netns`

**Comment by chetan-reddy** [comment]
With `pasta` now being the default on rootless podman, port forwarding works out of the box with gvisor's `network=sandbox` . Thanks @sbrivio-rh . The only caveat is that ports forwarded from 127.0.0.1 won't work because of pasta's 'fast bypass path for local connections', but you can always just connect locally using a public interface IP. Going to close this ticket.
